DM-19988: support code (and removal of dead code) for QuantumGraph generation rewrite

TallJimbo opened the pull request.

TallJimbo committed:
Replace QueryBuilder used in QuantumGraph generation.

MultipleDatasetQueryBuilder and its Registry-level interface were highly
specialized for QuantumGraph generation but didn't actually do what we need.
DataIdQueryBuilder is simpler, lower-level and hopefully more generally useful.
The higher-level logic for QuantumGraph generation is being moved to pipe_base,
which will now use DataIdQueryBuilder (and SingleDatasetQueryBuilder) directly.

TallJimbo committed:
Add optional data ID to Quantum (in Python).

Eventually we may want to add this to the database representation as
well, but that's a bit tricky due to the variable number of dimensions,
and if we partition the dataset table in the future that may provide
a model.

TallJimbo committed:
Add custom dictionary class for keys with "name" attributes.

TallJimbo committed:
Remove Registry methods for dealing with Quantum objects.

These weren't really thought through and are not currently in use,
so right now they're just a maintenance burden.

TallJimbo committed:
Add SQL-compiling __str__ to QueryBuilder.

TallJimbo committed:
Find the best join table for indirection, not just the first one.

TallJimbo committed:
Add matches method to DataId.

TallJimbo committed:
Augment and improve Quantum attributes.

TallJimbo committed:
Add property to get empty DimensionGraph for a DimensionUniverse.

TallJimbo committed:
Fix typo in docs.

TallJimbo committed:
Add temporal join between visit and calibration_label.

andy-slac commented on config/schema.yaml:
```diff
@@ -1030,6 +1030,66 @@ schema:
+        primary_key: true
+        nullable: false
+        doc: >
+          Unique (with instrument) integer identifier for an visit.
```
a visit

andy-slac commented on python/lsst/daf/butler/core/quantum.py:
```diff
@@ -21,49 +21,88 @@
+        The dimension values that identify this `Quantum`.
+    run : `Run`, optional
         The Run this Quantum is a part of.
+    initInputs : collection of `DatasetRef`
```
`optional`?

andy-slac commented on python/lsst/daf/butler/core/quantum.py:
```diff
@@ -21,49 +21,88 @@
+                 **kwargs):
+        super().__init__(**kwargs)
+        if taskClass is not None:
+            taskName = f"{taskClass.__module__}.{taskClass.__name__}"
```
Everyone seem to prefer f-strings for simple string operations, I still like old fashioned

    taskClass.__module__ + "." + taskClass.__name__

TallJimbo replied:
Yeah, I do like the f-string version better.  In this case there isn't much of a difference, but I like using the same pattern for simple and more complex cases that seem to be part of the same family of patterns, and I think this qualifies.

andy-slac commented on python/lsst/daf/butler/core/quantum.py:
```diff
@@ -21,49 +21,88 @@
+        if initInputs is None:
+            initInputs = {}
+        elif not hasattr(initInputs, "keys"):
+            initInputs = {ref.datasetType: ref for ref in initInputs}
```
Could it be that there is more than one `ref` in `initInputs` with the same dataset type?

TallJimbo replied:
No, the `PipelineTask` declarations already assume all `initInput` and `initOutput` dataset types are scalar (and they have to be because they have empty data IDs).

andy-slac commented on python/lsst/daf/butler/core/quantum.py:
```diff
@@ -21,49 +21,88 @@
+        self._initInputs = NamedKeyDict(initInputs)
+        self._predictedInputs = NamedKeyDict(predictedInputs if predictedInputs is not None else {})
+        self._actualInputs = NamedKeyDict(actualInputs if actualInputs is not None else {})
+        self._outputs = NamedKeyDict(outputs if outputs is not None else {})
```
Some serious typing could be saved if `NamedKeyDict` constructor accepted `None`. Or can't you use old pattern `outputs or {}`?

TallJimbo replied:
I don't think `None` makes sense for `NamedKeyDict` in general, so I'd prefer not to move the handling there.  And I once asked about a variant of the `outputs or {}` pattern on Slack/`#software-def`, and it was not popular (people considered it a Perl pattern that was a Python antipattern).  But I've come up with another solution: it's safe to pass an empty tuple `NamedKeyDict`'s constructor, so I'll just use that as the default for these arguments.

andy-slac reviewed (approved):
Looks OK, few minor comments.

TallJimbo committed:
Fix doc errors in QueryBuilder.

TallJimbo committed:
Add QueryBuilder method for DataId-based WHERE expressions.

TallJimbo committed:
Make SqlRegistryDatabaseDict transaction usage mirror SqlRegistry's.

In particular, do not begin transactions before read-only operations,
and use nested transactions on write operations.

This seems to be necessary in order to do a Datastore read
inside a QueryBuilder result-iteration loop.

andy-slac commented on python/lsst/daf/butler/sql/queryBuilder.py:
```diff
@@ -260,12 +260,12 @@ def whereSqlExpression(self, sqlExpression, op=and_):
         ----------
-        sqlExpression
-            SQLAlchemy boolean column expression.
+        expression : str
```
backticks around str?

andy-slac reviewed (approved):
Looks OK, one minor comment
